File: solutions/cpp/food-chain/1/food_chain.cpp
// Dedicated to Shree DR.MDD – the force behind every triumph
#include <sstream>
#include <vector>
#include "food_chain.h"
using namespace std;

namespace food_chain {

  struct Part {
    const string creature;
    const string reaction;
    const string reason;
    const bool is_end;
  };

  static const vector<Part> data = vector<Part>{
    { "fly", "", "I don't know why she swallowed the fly. Perhaps she'll die.", false },
    { "spider", "It wriggled and jiggled and tickled inside her.", "", false },
    { "bird", "How absurd to swallow a bird!", "She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.", false },
    { "cat", "Imagine that, to swallow a cat!", "", false },
    { "dog", "What a hog, to swallow a dog!", "", false },
    { "goat", "Just opened her throat and swallowed a goat!", "", false },
    { "cow", "I don't know how she swallowed a cow!", "", false },
    { "horse", "She's dead, of course!", "", true }
  };

  string verse(int position) {
    position -= 1;
    stringstream song;
    song << "I know an old lady who swallowed a " << data[position].creature << "." << endl;
    if (!data[position].reaction.empty()) {
      song << data[position].reaction << endl;
    }
    if (!data[position].is_end) {
      for (int j = position; j > 0; j--) {
        if (data[j].reason.empty()) {
          song << "She swallowed the " << data[j].creature << " to catch the " << data[j - 1].creature << "." << endl;
        } else {
          song << data[j].reason << endl;
        }
      }
      song << data[0].reason << endl;
    }
    return song.str();
  }

  string verses(int begin, int finish) {
    stringstream lyrics;
    for (int k = begin; k <= finish; ++k) {
      lyrics << verse(k) << endl;
    }
    return lyrics.str();
  }

  string sing() {
    return verses(1, data.size());
  }

}

File: solutions/swift/food-chain/1/Sources/FoodChain/FoodChain.swift
struct FoodChain {
    private static let chain = ["fly", "spider", "bird", "cat", "dog", "goat", "cow", "horse"]

    static func song(start: Int, end: Int) -> String {
        let verses = (start...end).map { number in
            let animal = chain[number - 1]

            if animal == "fly" {
                return beginning(for: "fly")
            }

            if animal == "horse" {
                return beginning(for: "horse")
            }

            return beginning(for: animal) + litany(from: number - 1)
        }

        return verses.joined(separator: "\n\n")
    }

    private static func beginning(for animal: String) -> String {
        let verse = reaction(for: animal)

        if animal == "spider" {
            return intro(for: "spider") + "It " + verse
        }

        return intro(for: animal) + verse
    }

    private static func intro(for animal: String) -> String {
        "I know an old lady who swallowed a \(animal).\n"
    }

    private static func reaction(for animal: String) -> String {
        switch animal {
        case "spider": return "wriggled and jiggled and tickled inside her.\n"
        case "bird": return "How absurd to swallow a bird!\n"
        case "cat": return "Imagine that, to swallow a cat!\n"
        case "dog": return "What a hog, to swallow a dog!\n"
        case "goat": return "Just opened her throat and swallowed a goat!\n"
        case "cow": return "I don't know how she swallowed a cow!\n"
        case "horse": return "She's dead, of course!"
        default: return "I don't know why she swallowed the fly. Perhaps she'll die."
        }
    }

    private static func litany(from index: Int) -> String {
        let lines = stride(from: index, to: 0, by: -1).map { current in
            let currentAnimal = chain[current]
            let previousAnimal = chain[current - 1]

            let line = "She swallowed the \(currentAnimal) to catch the \(previousAnimal)"

            if previousAnimal == "spider" {
                return line + " that " + reaction(for: "spider")
            }

            if previousAnimal == "fly" {
                return line + ".\n" + reaction(for: "fly")
            }

            return line + ".\n"
        }

        return lines.joined()
    }
}